File: models_py/m4_framing.py
"""M4 - catcher framing (Python side of the parity gate).

Logistic GLM of called-strike probability on pitch location. Called strikes above the
location-expected rate, summed per catcher and valued at ~0.125 runs/strike, are framing runs.
Reads ONLY the gold feed; writes a flat result file.
"""
from __future__ import annotations

import json
import os

import pandas as pd
import statsmodels.api as sm

RUN_PER_STRIKE = 0.125
MIN_PITCHES = 2000
FEATURES = ["plate_x", "plate_z", "plate_x_sq", "plate_z_sq"]


def main() -> None:
    gold = os.getenv("MLB_GOLD_DIR", "data/gold")
    res = os.getenv("MLB_RESULTS_DIR", "data/results")
    os.makedirs(res, exist_ok=True)

    df = pd.read_parquet(f"{gold}/called_pitches.parquet")
    x = sm.add_constant(df[FEATURES].to_numpy())
    y = df["is_called_strike"].to_numpy()
    model = sm.GLM(y, x, family=sm.families.Binomial()).fit()  # IRLS, matches R glm

    df["pred"] = model.predict(x)
    df["framing"] = (df["is_called_strike"] - df["pred"]) * RUN_PER_STRIKE
    cs = df.groupby(["catcher_id", "season"]).agg(
        runs=("framing", "sum"), pitches=("framing", "size")
    ).reset_index()
    q = cs[cs["pitches"] >= MIN_PITCHES]

    names = ["intercept", *FEATURES]
    out = {
        "n": int(len(df)),
        "coef": {n: round(float(c), 8) for n, c in zip(names, model.params, strict=True)},
        "top_framer_runs": round(float(q["runs"].max()), 4),
        "bottom_framer_runs": round(float(q["runs"].min()), 4),
        "n_qualified": int(len(q)),
    }
    with open(f"{res}/m4_framing__py.json", "w") as f:
        json.dump(out, f, indent=2)
    print(json.dumps({"stage": "models_py", "model": "m4_framing", **out}))


if __name__ == "__main__":
    main()

File: models_r/m4_framing.R
# M4 - catcher framing (R side of the parity gate).
# Logistic GLM of called-strike probability on pitch location; called strikes above expected,
# valued at ~0.125 runs/strike and summed per catcher, are framing runs.
suppressMessages({
  library(arrow)
  library(dplyr)
  library(jsonlite)
})

RUN_PER_STRIKE <- 0.125
MIN_PITCHES <- 2000

gold <- Sys.getenv("MLB_GOLD_DIR", "data/gold")
res <- Sys.getenv("MLB_RESULTS_DIR", "data/results")
dir.create(res, showWarnings = FALSE, recursive = TRUE)

df <- read_parquet(file.path(gold, "called_pitches.parquet"))
m <- glm(is_called_strike ~ plate_x + plate_z + plate_x_sq + plate_z_sq,
         data = df, family = binomial())

df$pred <- predict(m, type = "response")
df$framing <- (df$is_called_strike - df$pred) * RUN_PER_STRIKE
cs <- df %>%
  group_by(catcher_id, season) %>%
  summarise(runs = sum(framing), pitches = n(), .groups = "drop") %>%
  filter(pitches >= MIN_PITCHES)

co <- coef(m)
out <- list(
  n = nrow(df),
  coef = list(
    intercept = round(co[["(Intercept)"]], 8),
    plate_x = round(co[["plate_x"]], 8),
    plate_z = round(co[["plate_z"]], 8),
    plate_x_sq = round(co[["plate_x_sq"]], 8),
    plate_z_sq = round(co[["plate_z_sq"]], 8)
  ),
  top_framer_runs = round(max(cs$runs), 4),
  bottom_framer_runs = round(min(cs$runs), 4),
  n_qualified = nrow(cs)
)

# digits = 12: preserve precision for EXACT parity (see DECISIONS.md).
write_json(out, file.path(res, "m4_framing__r.json"), auto_unbox = TRUE, pretty = TRUE, digits = 12)
cat(toJSON(c(list(stage = "models_r", model = "m4_framing"), out), auto_unbox = TRUE, digits = 12), "\n")
